### PR TITLE
feat(events): complex event pattern matching with the `Match` class

### PR DESCRIPTION
**Pull request by otaviomacedo**
The current API restricts event pattern matchers to arrays of strings, which do not cover all the functionality available for EventBridge patterns.

Add a new matcher API that gets translated to the low-level JSON pattern matchers, to be consumed by EventBridge. This high-level API also adds:

* **Better ergonomics**: users don’t have to remember where to use curly or square brackets; 'OR' and 'AND' conditions (where applicable) are clearly indicated.
* **Aliases/syntax sugar**: `ipAddressRange()`, `anythingButPrefix()` and `interval()` hide some of the lower level complexity.
* **Validation**: you can’t pass an invalid IP address range, or an interval with lower and upper bounds swapped, or empty lists of conditions.

Resolves https://github.com/aws/aws-cdk/issues/20486, resolves https://github.com/aws/aws-cdk/issues/13330.

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-events/lib/event-pattern.ts [hunk 1]
```diff
@@ -37,15 +349,15 @@ export interface EventPattern {
    *
    * @default - No filtering on version
    */
-  readonly version?: string[];
+  readonly version?: string[] | IMatcher;
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events/lib/event-pattern.ts:352]: Avoid using union types in L2 libraries. Even though they work fine in TypeScript, they do not behave well in many of the other languages we target (See [Design Guidelines, API Design, Props, Unions](https://github.com/aws/aws-cdk/blob/v1-main/docs/DESIGN_GUIDELINES.md#unions)).

File: packages/@aws-cdk/aws-events/lib/event-pattern.ts [hunk 2]
```diff
@@ -1,3 +1,315 @@
...
+
+  toEventBridgeMatcher(): any[] {
+    return this.matchers.flatMap(m => m.toEventBridgeMatcher());
+  }
+}
+
+/**
+ * A collection of matchers for constructing event patterns
+ */
+export class Matchers {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events/lib/event-pattern.ts:185]: I think I'd prefer `Match` as a class name? It'll look nicer in code? As in, `Match.anythingBut(...)` ?

**Review by rix0rrr — CHANGES_REQUESTED** [review]
Nice! I love this! Just need to find a different way to encode the matchers... via tokens maybe?

**Comment by mrgrain** [comment]
@otaviomacedo Is this missing a Matcher for an AND condition? Or is that something the service doesn't support?

**Comment by otaviomacedo** [comment]
> @otaviomacedo Is this missing a Matcher for an AND condition? Or is that something the service doesn't support?

There is no general way to represent and AND condition. But there are two special cases:
1. The list of properties is interpreted as a conjunction: `{"foo": ["a"], bar: ["b"]}` means that you expect both fields to match.
2. In numeric patterns, the list of conditions passed to the array is interpreted as conjunction: `foo: [{"numeric": [">", 0, "<", 5]}]` means that the number should be greater than 0 _and_ less than 5. This is exposed by the `Match.allOf()` method.

**Review by rix0rrr — APPROVED** [review]

**Pull request merged**